### ISSUE TITLE
Reintroduce the ability to replace component (custom element's) tag with components template

### ISSUE THREAD

**Issue by brgrz** [angular/angular#25748]
Since Angular 2 shipped we were faced with the decision of the Angular team that the replace functionality in directives is gone.

Over the years we kinda accepted that but personally (and lots of other devs too) I resorted to hacks like the [removeHost] (https://stackoverflow.com/questions/34280475/remove-the-host-html-element-selectors-created-by-angular-component) directive to get rid of the extra custom element tags all over the generated markup.

The problem of extra tags in my opinion is only getting worse because of recent adoptions of both CSS Flexbox and lately CSS Grid. Both require working with immediate children in our markup and extra levels of element tags brought in by the use of components (which I agree is a great concept) are making the use of both new CSS features really PITA.

The issue is not solved by using attribute directives (components with [myComponent] selector) and applied to placeholder divs because those divs too are messing with markup and CSS Flexbox and Grid.

This issue can be solved to some degree by applying classes to our component references like <myComponent class="grid something"> but in reality is not feasible or maintainable in the long run.

There is also another issue that exists and it has to do with the rendering of the components that act as a UI layout (master pages) components and are tied to the routes in routing module and those cannot have classes applied (this issue sums up the problem https://stackoverflow.com/questions/45414515/apply-class-to-every-routed-component-in-angular).

I was just wondering if anyone on the Angular team recognizes the existance of these issues and if there are any plans to solve them (possibly by bringing back the replace functionality or some other way) because it seems to give headaches to lots of devs around the globe.

Best, M.

**Comment by dawidgarus** [comment]
It might be good idea to allow using components on `<ng-container>` like this: `<ng-container myComponent></ng-container>`.

**Comment by brgrz** [comment]
@dawidgarus certainly this would help

**Comment by dawidgarus** [comment]
@brgrz as a workaround (ugly) You can do something like this: https://stackblitz.com/edit/angular-xcse9s?file=src%2Fapp%2Fhello.component.ts - wrap component's template in ng-template, render it next to host element and detach host element.

**Comment by pkozlowski-opensource** [comment]
@brgrz thnx for opening this issue and a good summary of the issues with the current model. We are painfully aware of the limitations but host-less components has their problems too. One of the questions I was banging my head against is this: what should happen with a hose-less component which has bindings / event listeners / directives on it?

Let's consider this example:

```typescript
@Component({
  selector: 'my-cmpt',
  template: `
    <button>first</button>
    <button>second</button> 
  `
})
class MyComponent {
}
```

and its usage: `<my-component (click)="doSth()" ngbTooltip="I'm a tooltip">`.

If we introduce host-less components we must be able to answer following questions:
* what should happen with the `click` handler? 
* what should happen with the tooltip directive?

Those questions got precise and definitive answers in the current model. 

There are obviously more issues that need careful design and planning (ex. relation to custom elements) but for now I would like to better understand how the "replace" option is different from using an attribute selector and what practical issues you are encountering. A real-life example in stackblitz would help get the discussion going.

@dawidgarus your proposal suffers from the same issues regarding event handlers / directives as described above.

In short: we are very much aware of the general problem but:
* lacking more detailed description / examples of things that are causing grief in practice - please comment here _with stackblitz_;
* need to come up with a design that "works" and is correct / self-consistent.

**Comment by dawidgarus** [comment]
@pkozlowski-opensource I don't think there are any issues with ng-container approach. You can't bind to events with vanilla ng-container and You still wouldn't be able to do it with ng-container with component. Same thing with directives. You can create directive on ng-container and You still will be able.

**Comment by brgrz** [comment]
@pkozlowski-opensource the issue manifest itself when when our components are layout components (so called master pages with various router outlets) and those layout components are also flex or grid containers (display: grid, display: flex). This gets worse when nesting such components.

The introduction of extra elements both at the level of those components and at the level of their children (which may also be components) messes with CSS styles being applied correctly. 

And it increases the coupling between markup and CSS because we then have to target those elements either by their names or apply additional classes to them.

This happens with many CSS frameworks out there, for instance Zurb Foundation or when using plain CSS and the display: flex and display: grid features. 

Attribute directives do not solve this because the only difference is that there is now not a custom element rendered but a placeholder div with an attribute for directive - the issue is not in the type of the element (being custom or div) but rather in the existance of the element itself.

One workaround we use (besides the [removeHost] hack) is to move certain classes to the component host like so

```
import { Component, OnInit, HostBinding } from '@angular/core';

@Component({
	selector: 'proj-layout',
	templateUrl: './layout.component.html',
	styleUrls: ['./layout.component.css']
})
export class LayoutComponent implements OnInit {

	@HostBinding('attr.class') _class = 'grid-y medium-grid-frame'; // here we apply classes to the host element but we'd much rather have those classes within our markup
	constructor() { }

	ngOnInit() {
	}

}
```

The markup for this layout component looks like this

```
<div class="cell">
	<proj-header></proj-header>
</div>
<div class="cell auto medium-cell-block-container">
	<router-outlet></router-outlet>
</div>
```

but what we would like better is to have the container in the markup like so

```
<div class="grid-y medium-grid-frame">
   <div class="cell">
	<proj-header></proj-header>
   </div>
   <div class="cell auto medium-cell-block-container">
	<router-outlet></router-outlet>
   </div>
</div>
```

which of course we can, the only issue then is that there will be an extra useless component element rendered around it (using the HostAttribute solution we can actually use the host element as our container meaning it does not goes unused).

HTH

**Comment by brgrz** [comment]
@dawidgarus we are currently using the HostAttribute workaround but it is not ideal because we don't to bring the styling into the component class however this is one way of solving it... thanks for the stackblitz suggestion

**Comment by trotyl** [comment]
Duplicate of https://github.com/angular/angular/issues/18877, https://github.com/angular/angular/issues/22208

**Comment by brgrz** [comment]
@trotyl as I thought the pain was there for frontend devs and designers for quite some time..

Do we know if there is an official Angular solution/feature coming anytime soon?
